src/journal/bsod.c

```diff
@@ -0,0 +1,244 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to seek to start of jornal: %m");
+
+        r = sd_journal_next(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read next journal entry: %m");
+        if (r == 0) {
+                log_debug("No entries in the journal");
+                return r;
+        }
+
+        r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = memdup_suffix0((const char*)d + STRLEN("MESSAGE="), l - STRLEN("MESSAGE="));
+        if (! message)
+                return log_oom();
+
+        *ret = TAKE_PTR(message);
+
+        return r;
+}
+
+static int find_next_free_vt(int fd, int *ret_free_vt, int *ret_original_vt) {
+        assert(fd);
+
+        size_t i;
+        struct vt_stat terminal_status;
+
+        if (ioctl(fd, VT_GETSTATE, &terminal_status) < 0)
+                return -errno;
+
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++)
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+
+        *ret_free_vt = i;
+        *ret_original_vt = terminal_status.v_active;
+
+        return 0;
+}
+
+static int display_emergency_message_fullscreen(char * message) {
+        assert(message);
+        int r, free_vt, original_vt, qr_code_start_row = 1, qr_code_start_column = 1;
+        struct winsize w;
+        char tty[STRLEN("/dev/tty") + DECIMAL_STR_MAX(int) + 1];
+        _cleanup_close_ int fd = -EBADF;
+        _cleanup_fclose_ FILE *stream = NULL, *out = NULL;
+        char read_character_buffer = '\0';
+
+        fd = open_terminal("/dev/tty1", O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return log_error_errno(fd, "Failed to open tty1: %m");
+
+        r = find_next_free_vt(fd, &free_vt, &original_vt);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find a free VT: %m");
+
+        xsprintf(tty, "/dev/tty%d", free_vt + 1);
+
+        fd = open_terminal(tty, O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return log_error_errno(fd, "Failed to open tty: %m");
+
+        if (ioctl(fd, TIOCGWINSZ, &w) < 0)
+                log_warning_errno(errno, "Failed to fetch tty size, ignoring: %m");
+
+        if (ioctl(fd, VT_ACTIVATE, free_vt + 1) < 0)
+                return log_error_errno(errno, "Failed to activate tty: %m");
+
+        r = loop_write(fd, ANSI_BACKGROUND_BLUE, strlen(ANSI_BACKGROUND_BLUE), false);
+        if (r < 0)
+                log_warning_errno(r, "Failed to change terminal color to blue, ignoring: %m");
+
+        r = loop_write(fd, ANSI_HOME_CLEAR, strlen(ANSI_HOME_CLEAR), false);
+        if (r < 0)
+                log_warning_errno(r, "Failed to clear terminal, ignoring: %m");
+
+        qr_code_start_row = (int) (w.ws_row * 0.75);
+        qr_code_start_column = (int) (w.ws_col * 0.75);
```

Comment:
if TIOCGWINSZ failed you'll access uninitialized memory here. maybe just initialize `w` to 80x25 in case of failure, i.e. the traditional VGA text console size from the 1990s.

i usually would avoid floating point calculations if we don't need to. To take 3/4th of something just write it out as integer calculations:

`w.ws_row * 3U / 4U`